### ISSUE TITLE
Remove unnecessary null pointer checks

### ISSUE THREAD

**Issue by elfring** [godotengine/godot#55]
[An extra null pointer check is not needed in functions](http://dietmar-kuehl.de/mirror/c++-faq/freestore-mgmt.html#faq-16.8 "Do I need to check for NULL before delete p?") like the following.
* [OSNacl::finalize_core](https://github.com/okamstudio/godot/blob/0b806ee0fc9097fa7bda7ac0109191c9c5e0a1ac/platform/nacl/os_nacl.cpp#L143 "OSNacl::finalize_core function")
* [OS_Unix::finalize_core](https://github.com/okamstudio/godot/blob/0b806ee0fc9097fa7bda7ac0109191c9c5e0a1ac/drivers/unix/os_unix.cpp#L127 "OS_Unix::finalize_core function")
* [OS_Windows::finalize_core](https://github.com/okamstudio/godot/blob/0b806ee0fc9097fa7bda7ac0109191c9c5e0a1ac/platform/windows/os_windows.cpp#L1123 "OS_Windows::finalize_core function")

**Comment by akien-mga** [comment]
I guess the fix is simply to drop the ifs? Does it apply only for ```delete``` or also  ```memdelete```? Feel free to propose a pull request to fix it.

**Comment by trollworkout** [comment]
@elfring  @akien-mga  I did some investigation and it seems most people recommend doing null check anyway. 

If you are trying to delete a null pointer is likely you might have a bug somewhere. Also say you don't check for null you may have a bug that will spam delete on many null pointers affecting performance and you will never notice it without a check.

**Comment by elfring** [comment]
* @trollworkout: Where do you see recommendations for [extra null pointer checks before calling the delete operator](https://isocpp.org/wiki/faq/freestore-mgmt#delete-handles-null "Do I need to check for NULL before delete p?")?
* @touilleMan: Thanks for [your small source code improvement](https://github.com/godotengine/godot/commit/cabf9234849508ba7285d1a5bafd3bab692eb80f "Deletion of a few unnecessary checks").

**Comment by trollworkout** [comment]
@elfring  I just simply researched online for opinions, forum threads, reddit posts etc from C++ devs based on their experience with the said issue. I formed my own post comment and opinion based on what experienced C++ devs have posted.

for example http://stackoverflow.com/questions/615355/is-there-any-reason-to-check-for-a-null-pointer-before-deleting

The consensus while not required to do null checks before delete it is useful to do it anyway because trying to delete null pointers may be a bug sign.

**Comment by elfring** [comment]
1. How useful is it really to repeat unnecessary [null pointer checks](http://stackoverflow.com/questions/615355/is-there-any-reason-to-check-for-a-null-pointer-before-deleting "Is there any reason to check for a NULL pointer before deleting?")?
2. Can it be [a class like "std::unique_ptr"](https://isocpp.org/wiki/faq/cpp11-library "Description for useful C++ classes") should be reused more often nowadays (instead of even calling the delete operator explicitly)?

**Comment by trollworkout** [comment]
@elfring  Well they say the issue is delete simply does nothing if the pointer is NULL. The issue is delete returns nothing. Sending a null pointer to delete may be a symptom that your code is bad somewhere. So you do the check for null to catch a possible flaw in the code.

**Comment by elfring** [comment]
How do you think about to improve the affected error detection by [reusing appropriate C++ exceptions](https://isocpp.org/wiki/faq/exceptions "Why use exceptions?") instead of leaving questionable checks over?

**Comment by akien-mga** [comment]
> Sending a null pointer to delete may be a symptom that your code is bad somewhere. So you do the check for NULL to catch a possible flaw in the code.

But that's not the case here. There was no error message raised in the case of a null pointer, only:
```
if (pointer)
    delete pointer;
```

We wouldn't have removed those extraneous checks if the code were:
```
if (pointer)
    delete pointer;
else
    print_line("Oh shi--- pointer was null, somebody messed up!);
```
;)

**Comment by trollworkout** [comment]
@elfring  see here 
http://www.cplusplus.com/reference/new/operator%20delete/

returns nothing

You could be sending null pointers to delete and would not even know it.  

C++ devs say is useful tool in preventing certain bugs from seeping in

**Comment by elfring** [comment]
@trollworkout: Are you looking for a customised debug memory management system?

**Comment by akien-mga** [comment]
> You could be sending null pointers to delete and would not even know it.
>
> C++ devs say is useful tool in preventing certain bugs from seeping in

In the stackoverflow topic you linked, those "C++ devs" were actually saying that such checks are not needed. Then some added that in some specific cases, they can be useful *if* one needs to intercept the fact that the pointer was null when an attempt to delete it was made. It does *not* mean that we're in this case, and in this same thread there is a clear consensus that checking for null pointer before deleting in C++ is *not* needed generally.

**Comment by trollworkout** [comment]
@elfring i don't know :)

I was reading about the issue and found most people do the checks in their own application because sending null pointers to delete may be a sign of bugs. So is more of a tool like you said 

if (pointer)
    delete pointer;
else if (debugmode)
    print_line("you're trying to delete a null pointer);

Is not necessary but may be an useful tool

**Comment by trollworkout** [comment]
@akien-mga  correct :)

**Comment by akien-mga** [comment]
Fixed by #3230.
